### ISSUE TITLE
Flask-SocketIO novice with a couple KeyErrors

### ISSUE THREAD

**Issue by killerHELIX** [miguelgrinberg/Flask-SocketIO#486]
Hi,

I'm getting a KeyError every time I emit from Javascript to my server.py.  The internet is a great tool, but hasn't been able to help me.  Check out the relevant code below (forgive the mess):

app.js
```
var app = angular.module('app', []);
var socket = io.connect('https://' + document.domain + ':' + location.port);

app.controller('owner', function($scope, $sce, shared) {

    socket.on('connect', function() {
        console.log('Owner CTRL connected.');
        console.log('$scope.nowPlaying.url: ' + $scope.nowPlaying.url);
        console.log("path: " + $scope.nowPlayingSrc);
        console.log("audio path: " + document.getElementById('nowPlaying').src);

    });

    // a debug function that currently is throwing keyErrors
    $scope.debugPopulateDB = function() {
        console.log("Entered debugPopulateDB on owner CTRL (app.js)");
        socket.emit('debugPopulateDB')    
    };
    
    // pushes selected track to queue (also throws keyError)
    $scope.push = function() {
        console.log("Entered $scope.addToQueue on owner CTRL (app.js)");
        socket.emit('findTrack', $scope.selectedTrack);
    };
});

```

server.py
```
import os
import sys
from flask_socketio import SocketIO, emit
from flask import Flask, render_template, request, session, redirect

app = Flask(__name__)
app.config['SECRET_KEY'] = os.urandom(24).encode('hex')

# prevents UnicodeDecodeError
reload(sys)
sys.setdefaultencoding('utf8')

socketio = SocketIO(app)

@socketio.on('connect')
def makeConnection():
    print('Connected.')
    
    print("Printing all tracks...")
    for track in tracks.find():
        pprint.pprint(track)
        
@socketio.on('findTrack')
def findTrack(track):
	print("Entered findTrack in server.py with param " + track)
	
@socketio.on('debugPopulateDB')
def debug():
	print("Entered debugPopulateDB in server.py")
	inserted_tracks = {}
	custom_index = ['A', 'B', 'C']
	
	# iterate horizontally
	for index in custom_index:
		for i in range(1,3):
			print(index + i)
    
@app.route('/', methods=['GET', 'POST'])
def renderIndex():
	
	# initialize session variables
	if 'user' not in session:
		session['user'] = None
		
	print(request.method)
	return render_template('index.html')

# start the server
if __name__ == '__main__':
    socketio.run(app, host=os.getenv('IP', '0.0.0.0'), port =int(os.getenv('PORT', 8080)), debug=True)
```

The Traceback I'm getting is below.
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Flask-0.12.2-py2.7.egg/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/Flask_SocketIO-2.8.6-py2.7.egg/flask_socketio/__init__.py", line 42, in __call__
    start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/socketio/server.py", line 353, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 289, in handle_request
    self.disconnect(sid)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 203, in disconnect
    del self.sockets[sid]
KeyError: '0909ebb81faa476581ea39ded047bcce'
```

Apologies if this isn't enough information.  I'm using Angular 1.4.8 in JS.

Also, I'm randomly getting a different KeyError. Haven't found a way to reproduce it.  Traceback:
```
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/Flask-0.12.2-py2.7.egg/flask/app.py", line 1997, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/Flask_SocketIO-2.8.6-py2.7.egg/flask_socketio/__init__.py", line 42, in __call__
    start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/socketio/server.py", line 353, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 268, in handle_request
    self.disconnect(sid)
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 202, in disconnect
    self._get_socket(sid).close()
  File "/usr/local/lib/python2.7/dist-packages/engineio/server.py", line 411, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'
```

**Comment by killerHELIX** [comment]
Should've spent more time on this issue before posting.   Because I was trying to concatenate a string and an integer, everything hit the fan.  Casting `i` to a str fixed the issue.  Still getting that random KeyError though.

**Comment by miguelgrinberg** [comment]
Try upgrading the package python-engineio to the latest release. There was a bug in older versions of this package that made errors in your application be incorrectly reported as these KeyErrors you show above.

**Comment by miguelgrinberg** [comment]
I'm going to assume this was addressed by the 1.6.0 release of python-engineio. If you have any remaining problems, feel free to reopen this issue or open a new one.
